apps/system/lockscreen/js/widgets/clock/lockscreen_clock_widget.js

```diff
@@ -50,6 +52,56 @@
     this.logger.debug('Clock updated', now);
   };
 
+  LockScreenClockWidget.prototype.updateAlarm =
+	function() {
+    this.fetchAlarmData()
+    .then((alarmData) => {
+      // if there is no alarm
+      if(!alarmData) {
+        this.resources.elements.alarm.classList.add('no-alarms');
+        this.resources.elements.alarm.classList.remove('has-alarm');
+        return;
+      }
+
+      var type = 'AM',
+          hour = parseInt(alarmData.hour, 10);
+
+      // Decide wheter to use AM or PM
+      if(hour > 12) {
+        type = 'PM';
+        hour = hour - 12;
+      }
+
+      var fullAlarmTime = hour + ':' + alarmData.minute + type;
+      this.resources.elements.alarmtime.textContent = fullAlarmTime;
+      this.resources.elements.alarm.classList.remove('no-alarms');
+      this.resources.elements.alarm.classList.add('has-alarm');
+      this.logger.debug('Alarm updated to :', fullAlarmTime);
+    });
+	};
+
+  /*
+   * Gets the Alarm Data from the DataStore.
+   */
+  LockScreenClockWidget.prototype.fetchAlarmData = function() {
+    return navigator.getDataStores('alarms')
+      .then((stores) => {
+        return [stores[0].getLength(), stores[0]];
+      }).then(([len, store]) => {
+        if(0 === len) {
+          return;
+        }
+        return store.get(1);
+      }).then((fetchedData = {}) => {
+        if(0 === Object.keys(fetchedData).length) {
+          return null;
+        }
+        return fetchedData;
+      }).catch(function(e){
+        console.error('Error occured while getting alarm data : ', e);
```

Comment:
Please note what I've commented: you need to re-throw the error if you want the following chains after this method (e.g. `fetchAlarmData().then()`) can be interrupted as well. It's important in common cases since the following chains may do some effects or even unexpected effects while without a correct ascendants. Therefore if you thought it's not what you want, and you silent the error here **intentionally**, please comment it here. Or, please re-throw it after you log the error.

And why can't we just leave the error to the following one without dealing with it here, is because you never know who may concate your Promised methods after 3 months, although currently they're all only maintained and designed by you. So it's a defensive design: even others may screw the code by concating a bad Promise after this method, you can still see the error logs in the console, and you're sure if there are any error in your chain, the following Promises wouldn't get any chance to execute, unless they intentionally silent it by catching it (so the next reviewer should ask them the reason to do that like me).